File: ViewProfit/table_benchmarks.py
# -*- coding: utf-8 -*-


import numpy as np
from PySide2.QtCharts import QtCharts
from PySide2.QtCore import QDateTime, Qt, Signal
from PySide2.QtSql import QSqlTableModel

from ViewProfit.model_benchmark import ModelBenchmark
from ViewProfit.table_base_ib import TableBaseIB


class TableBenchmarks(TableBaseIB):
    new_mouse_coords = Signal(object,)

    def __init__(self, app, name):
        TableBaseIB.__init__(self, app, name)

        self.model = ModelBenchmark(self, app.db)

        self.init_model()

        self.lineedit_name.setText(name)

    def init_model(self):
        self.model.setTable(self.name)
        self.model.setEditStrategy(QSqlTableModel.OnManualSubmit)
        self.model.setSort(1, Qt.SortOrder.DescendingOrder)
        self.model.setHeaderData(1, Qt.Horizontal, "Date")
        self.model.setHeaderData(2, Qt.Horizontal, "Monthly Value %")
        self.model.setHeaderData(3, Qt.Horizontal, "Accumulated %")

        self.model.select()

        self.table_view.setModel(self.model)
        self.table_view.setColumnHidden(0, True)  # do no show the id column

    def recalculate_columns(self):
        list_value = []

        for n in range(self.model.rowCount()):
            list_value.append(self.model.record(n).value("value"))

        if len(list_value) > 0:
            list_value.reverse()

            list_value = np.array(list_value) * 0.01

            accumulated = (np.cumprod(list_value + 1.0) - 1.0) * 100.0

            accumulated = accumulated[::-1]  # reversed array

            for n in range(self.model.rowCount()):
                rec = self.model.record(n)

                rec.setGenerated("accumulated", True)

                rec.setValue("accumulated", float(accumulated[n]))

                self.model.setRecord(n, rec)

    def show_chart(self):
        self.clear_charts()

        self.chart1.setTitle(self.name)
        self.chart2.setTitle(self.name)

        series0 = QtCharts.QLineSeries()
        series1 = QtCharts.QLineSeries()

        series0.setName("Monthly Value")
        series1.setName("Accumulated")

        series0.hovered.connect(self.on_hover)
        series1.hovered.connect(self.on_hover)

        for n in range(self.model.rowCount()):
            qdt = QDateTime.fromString(self.model.record(n).value("date"), "dd/MM/yyyy")

            epoch_in_ms = qdt.toMSecsSinceEpoch()

            v = self.model.record(n).value("value")
            accu = self.model.record(n).value("accumulated")

            series0.append(epoch_in_ms, v)
            series1.append(epoch_in_ms, accu)

        self.chart1.addSeries(series0)
        self.chart2.addSeries(series1)

        # add axes to series 0

        axis_x = QtCharts.QDateTimeAxis()
        axis_x.setTitleText("Date")
        axis_x.setFormat("dd/MM/yyyy")
        axis_x.setLabelsAngle(-10)

        axis_y = QtCharts.QValueAxis()
        axis_y.setTitleText("%")
        axis_y.setLabelFormat("%.2f")

        self.chart1.addAxis(axis_x, Qt.AlignBottom)
        self.chart1.addAxis(axis_y, Qt.AlignLeft)

        series0.attachAxis(axis_x)
        series0.attachAxis(axis_y)

        # add axes to series 1

        axis_x = QtCharts.QDateTimeAxis()
        axis_x.setTitleText("Date")
        axis_x.setFormat("dd/MM/yyyy")
        axis_x.setLabelsAngle(-10)

        axis_y = QtCharts.QValueAxis()
        axis_y.setTitleText("%")
        axis_y.setLabelFormat("%.2f")

        self.chart2.addAxis(axis_x, Qt.AlignBottom)
        self.chart2.addAxis(axis_y, Qt.AlignLeft)

        series1.attachAxis(axis_x)
        series1.attachAxis(axis_y)

File: src/table_benchmarks.cpp
#include "table_benchmarks.hpp"
#include <QSqlQuery>
#include "chart_funcs.hpp"

TableBenchmarks::TableBenchmarks(QWidget* parent) : TableBase(parent) {
  type = TableType::Benchmark;

  fund_cfg_frame->hide();

  radio_chart2->setText("Accumulated");
  radio_chart1->setText("Monthly Value");

  connect(spinbox_months, QOverload<int>::of(&QSpinBox::valueChanged), [&](int value) { show_chart(); });
}

void TableBenchmarks::init_model() {
  model->setTable(name);
  model->setEditStrategy(QSqlTableModel::OnManualSubmit);
  model->setSort(1, Qt::DescendingOrder);
  model->setHeaderData(1, Qt::Horizontal, "Date");
  model->setHeaderData(2, Qt::Horizontal, "Monthly Value %");
  model->setHeaderData(3, Qt::Horizontal, "Accumulated %");
  model->select();

  table_view->setModel(model);
  table_view->setColumnHidden(0, true);
}

void TableBenchmarks::calculate() {
  QVector<double> list_values;
  QVector<double> accu;

  for (int n = 0; n < model->rowCount(); n++) {
    list_values.append(model->record(n).value("value").toDouble());
  }

  if (!list_values.empty()) {
    std::reverse(list_values.begin(), list_values.end());

    for (auto& value : list_values) {
      value = value * 0.01 + 1.0;
    }

    // cumulative product

    accu.resize(list_values.size());

    std::partial_sum(list_values.begin(), list_values.end(), accu.begin(), std::multiplies<>());

    for (auto& value : accu) {
      value = (value - 1.0) * 100;
    }

    std::reverse(accu.begin(), accu.end());

    for (int n = 0; n < model->rowCount(); n++) {
      auto rec = model->record(n);

      rec.setGenerated("accumulated", true);

      rec.setValue("accumulated", accu[n]);

      model->setRecord(n, rec);
    }

    show_chart();
  }
}

void TableBenchmarks::show_chart() {
  clear_charts();

  chart1->setTitle(name.toUpper());
  chart2->setTitle(name.toUpper());

  add_axes_to_chart(chart1, "%");
  add_axes_to_chart(chart2, "%");

  auto s1 = add_series_to_chart(chart1, model, "Monthly Value", "value");

  connect(s1, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout1, s1->name()); });

  model->submitAll();

  QVector<int> dates;
  QVector<double> values;
  QVector<double> accu;

  auto query = QSqlQuery(db);

  query.prepare("select distinct date,value from " + name + " order by date desc");

  if (query.exec()) {
    while (query.next() && dates.size() < spinbox_months->value()) {
      dates.append(query.value(0).toInt());
      values.append(query.value(1).toDouble());
    }
  } else {
    qDebug("Failed to get inflation table values!");
  }

  if (dates.empty()) {
    qDebug("oi");
    return;
  }

  std::reverse(dates.begin(), dates.end());
  std::reverse(values.begin(), values.end());

  for (auto& v : values) {
    v = v * 0.01 + 1.0;
  }

  // cumulative product

  accu.resize(values.size());

  std::partial_sum(values.begin(), values.end(), accu.begin(), std::multiplies<>());

  for (auto& v : accu) {
    v = (v - 1.0) * 100;
  }

  auto s2 = add_series_to_chart(chart2, dates, accu, "Accumulated");

  connect(s2, &QLineSeries::hovered, this,
          [=](const QPointF& point, bool state) { on_chart_mouse_hover(point, state, callout2, s2->name()); });
}